lib/internal/http2/core.js

```diff
@@ -2696,8 +2696,12 @@ function connect(authority, options, listener) {
   session[kAuthority] = `${options.servername || host}:${port}`;
   session[kProtocol] = protocol;
 
-  if (typeof listener === 'function')
-    session.once('connect', listener);
+  if (typeof listener === 'function') {
+    if (session.connecting)
+      session.once('connect', listener);
+    else
+      Reflect.apply(listener, session, [socket]);
```

Comment:
@jasnell @mcollina Thinking about this a bit more, should we use `process.nextTick` and emit a `connect` event regardless rather than directly calling the `listener`? I know we don't recommend this code pattern anywhere but something like the following would still not work after this change:

```js
const session = http2.connect(SERVER_URL);
session.once('connect', listener);
```

Reply:
@apapirovski I think we should copy-paste what we are doing for http, https, net and tls. So, we should adhere to the prevailing pattern.

Reply:
I'm good with emitting on nextTick

Reply:
See f21670f5d51d451cd2f2203f14d0f804ba107905